libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs

```diff
@@ -514,70 +517,105 @@ protected async Task<DialogTurnResult> ContinueActionsAsync(DialogContext dc, ob
-                    if (ac != null && ac.Changes != null && ac.Changes.Count > 0)
+                    if (result.Status == DialogTurnStatus.CompleteAndWait)
                     {
-                        parentChanges = true;
+                        // Child dialog completed, but wants us to wait for a new activity
+                        result.Status = DialogTurnStatus.Waiting;
+                        return result;
                     }
 
-                    root = parent;
-                    parent = root.Parent;
-                }
+                    var parentChanges = false;
+                    DialogContext root = actionContext;
+                    var parent = actionContext.Parent;
+                    while (parent != null)
+                    {
+                        var ac = parent as ActionContext;
+                        if (ac != null && ac.Changes != null && ac.Changes.Count > 0)
+                        {
+                            parentChanges = true;
+                        }
 
-                // Execute next step
-                if (parentChanges)
-                {
-                    // Recursively call ContinueDialogAsync() to apply parent changes and continue
-                    // execution.
-                    return await root.ContinueDialogAsync(cancellationToken).ConfigureAwait(false);
-                }
+                        root = parent;
+                        parent = root.Parent;
+                    }
+
+                    // Execute next step
+                    if (parentChanges)
+                    {
+                        // Recursively call ContinueDialogAsync() to apply parent changes and continue
+                        // execution.
+                        return await root.ContinueDialogAsync(cancellationToken).ConfigureAwait(false);
+                    }
 
-                // Apply any local changes and fetch next action
-                await actionContext.ApplyChangesAsync(cancellationToken).ConfigureAwait(false);
-                actionDC = CreateChildContext(actionContext);
+                    // Apply any local changes and fetch next action
+                    await actionContext.ApplyChangesAsync(cancellationToken).ConfigureAwait(false);
+                    actionDC = CreateChildContext(actionContext);
+                }
+            }
+            finally
+            {
+                OnPopScopedServices(dc.Context);
             }
 
             return await OnEndOfActionsAsync(actionContext, cancellationToken).ConfigureAwait(false);
         }
 
+        /// <summary>
+        /// OnPushScopedServices provides ability to Push scoped services to the turnState.
+        /// </summary>
+        /// <remarks>If you override this you should make sure to call the base.OnPushScopedServices().</remarks>
+        /// <param name="turnContext">turnContext.</param>
+        protected virtual void OnPushScopedServices(ITurnContext turnContext)
```

Comment:
Scoped services as a concept is something that we should plan how to introduce to users. I'd love to understand how we can make the Dictionary with stacks intuitive to users without them understanding the internal implementation. We can chat about this in a call.

Reply:
I've been thinking about this and I think I understand what is going on. We'll need some kind of material to explain this to users at some point.

Reply:
we discussed thisoffline